neon_hana/streaming_client.py

```diff
@@ -0,0 +1,107 @@
+import io
+from asyncio import run
+from base64 import b64encode, b64decode
+from typing import Optional, Callable
+from mock.mock import Mock
+from threading import Thread
+from queue import Queue
+
+from ovos_dinkum_listener.voice_loop import DinkumVoiceLoop
+from ovos_dinkum_listener.voice_loop.hotwords import HotwordContainer
+from ovos_dinkum_listener.voice_loop.voice_loop import ChunkInfo
+from ovos_plugin_manager.templates.microphone import Microphone
+from ovos_plugin_manager.vad import OVOSVADFactory
+from ovos_utils.fakebus import FakeBus
+from speech_recognition import AudioData
+from ovos_utils import LOG
+from starlette.websockets import WebSocket
+
+
+class StreamMicrophone(Microphone):
+    def __init__(self, queue: Queue):
+        self.queue = queue
+
+    def start(self):
+        pass
+
+    def stop(self):
+        self.queue.put(None)
+
+    def read_chunk(self) -> Optional[bytes]:
+        return self.queue.get()
+
+
+class RemoteStreamHandler(Thread):
+    def __init__(self, mic: StreamMicrophone, session_id: str,
+                 input_audio_callback: Callable,
+                 client_socket: WebSocket,
+                 ww_callback: Callable, lang: str = "en-us"):
+        Thread.__init__(self)
+        self.session_id = session_id
+        self.ww_callback = ww_callback
+        self.input_audio_callback = input_audio_callback
+        self.client_socket = client_socket
+        self.bus = FakeBus()
+        self.mic = mic
+        self.lang = lang
+        self.hotwords = HotwordContainer(self.bus)
+        self.hotwords.load_hotword_engines()
+        self.vad = OVOSVADFactory.create()
+        self.voice_loop = DinkumVoiceLoop(mic=self.mic,
+                                          vad=self.vad,
+                                          hotwords=self.hotwords,
+                                          listenword_audio_callback=self.on_hotword,
+                                          hotword_audio_callback=self.on_hotword,
+                                          stopword_audio_callback=self.on_hotword,
+                                          wakeupword_audio_callback=self.on_hotword,
+                                          stt_audio_callback=self.on_input_audio,
+                                          stt=Mock(transcribe=Mock(return_value=[])),
+                                          fallback_stt=Mock(transcribe=Mock(return_value=[])),
```

Comment:
Consider `MagicMock`, which should stub out all of the necessary methods without having to be explicit

Reply:
Looks like the inferred return type is not a list if I use `MagicMock` (similar issue with transformers return value)
```
│ Exception in thread Thread-6:                                                                                                                                  │
│ Traceback (most recent call last):                                                                                                                             │
│   File "/usr/local/lib/python3.9/threading.py", line 980, in _bootstrap_inner                                                                                  │
│     self.run()                                                                                                                                                 │
│   File "/usr/local/lib/python3.9/site-packages/neon_hana/streaming_client.py", line 67, in run                                                                 │
│     self.voice_loop.run()                                                                                                                                      │
│   File "/usr/local/lib/python3.9/site-packages/ovos_dinkum_listener/voice_loop/voice_loop.py", line 269, in run                                                │
│     self._after_cmd(chunk)                                                                                                                                     │
│   File "/usr/local/lib/python3.9/site-packages/ovos_dinkum_listener/voice_loop/voice_loop.py", line 783, in _after_cmd                                         │
│     utts, stt_context = self._get_tx(stt_context)                                                                                                              │
│   File "/usr/local/lib/python3.9/site-packages/ovos_dinkum_listener/voice_loop/voice_loop.py", line 731, in _get_tx                                            │
│     filtered = [max(utts, key=lambda k: k[1])]                                                                                                                 │
│ ValueError: max() arg is an empty sequence 
```

Reply:
You can set a return type to anything you'd like with a `MagicMock`, but this is definitely not blocking:

```python
Python 3.12.6 (main, Sep  6 2024, 19:03:47) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from unittest.mock import MagicMock
>>>
>>> # Create a mock transcription result
>>> mock_transcription = [("Hello, this is a test", 0.9)]
>>>
>>> # Set up the STT objects using MagicMock
>>> stt = MagicMock()
>>> stt.transcribe.return_value = mock_transcription
>>>
>>> fallback_stt = MagicMock()
>>> fallback_stt.transcribe.return_value = mock_transcription
>>> stt.transcribe
<MagicMock name='mock.transcribe' id='4314005728'>
>>> stt.transcribe()
[('Hello, this is a test', 0.9)]
>>> fallback_stt.transcribe()
[('Hello, this is a test', 0.9)]
>>> type(fallback_stt.transcribe())
<class 'list'>
```